### PR TITLE
#1434 Added support for ZMSCORE new in Redis 6.2 RC

### PR DESCRIPTION
**Pull request by 2014BDuck**
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [X] Does `$ tox` pass with this change (including linting)?
- [X] Does travis tests pass with this change (enable it first in your forked repo and wait for the travis build to finish)?
- [X] Is the new or changed code fully tested?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?

_NOTE: these things are not required to open a PR and can be done
afterwards / while the PR is open._

### Description of change
Added support for zmscore command.
https://github.com/andymccurdy/redis-py/issues/1434
https://github.com/andymccurdy/redis-py/issues/1435

```
>>> r.zmscore(key="test", members=["member1", "member2", "invalid_member"])
[1.0, 2.0, None]
```

### CODE REVIEW

**Comment by adamserafini** [comment]
One comment: I think the new test introduced here never runs on CI because the base docker image for redis needs to bumped to something >= 6.2: https://github.com/andymccurdy/redis-py/blob/master/docker/base/Dockerfile#L1

so the the test is currently always skipped because `@skip_if_server_version_lt('6.1.240')` will be true.

**Comment by 2014BDuck** [comment]
> One comment: AFAICT, I think the new test introduced here never runs on CI because the base docker image for redis needs to bumped to something >= 6.2: https://github.com/andymccurdy/redis-py/blob/master/docker/base/Dockerfile#L1
> 
> The test is currently always skipped because `@skip_if_server_version_lt('6.1.240')` will be true.

Yes I see. So when I wrote the test, I removed the decorator and run it locally(tox) with version 6.2RC. 
But you are right I think we need to update the docker image thing for Codecov testing.

**Comment by chayim** [comment]
[This PR](https://github.com/andymccurdy/redis-py/pull/1536) does that. once it's merged we should re-run this flow.

File: redis/client.py [hunk 1]
```diff
@@ -3281,6 +3287,22 @@ def zunionstore(self, dest, keys, aggregate=None):
...
+        Returns the scores associated with the specified members
+        in the sorted set stored at key.
+
+        ``members`` should be a list of the member name.
+        Return type is a list of score.
+
+        If the member does not exist, a None will be returned
+        in corresponding position.
+        """
+        if not isinstance(members, list) or len(members) < 1:
```
Comment (chayim) [inline redis/client.py:3301]:
```suggestion
        if not members:
```

File: redis/client.py [hunk 2]
```diff
@@ -3281,6 +3287,22 @@ def zunionstore(self, dest, keys, aggregate=None):
         """
         return self._zaggregate('ZUNIONSTORE', dest, keys, aggregate)
 
+    def zmscore(self, key, members):
```
Comment (chayim) [inline redis/client.py:3290]:
```suggestion
    def zmscore(self, key, members=[]):
```

**Comment by chayim** [comment]
Hi @2014BDuck thanks for your submission! Everything passes with the current unit tests.

File: redis/client.py (hunk 2 above)
Comment (2014BDuck) [reply redis/client.py:3369]: Sorry for the late reply.

I am ok with a default value proposal (for compatible reason). But not sure if using a mutable type is fine. This contains risk that the default value might be modified.

So maybe I should change it with a default value None and modify the function body accordingly?

File: redis/client.py (hunk 1 above)
Comment (2014BDuck) [reply redis/client.py:3380]: So we should check if it's None or not and if it's a list type data?

**Comment by 2014BDuck** [comment]
@chayim Hi sorry for the late reply. I would prefer using an immutable type as a default value, what do you think?

File: redis/client.py (hunk 1 above)
Comment (chayim) [reply redis/client.py:3380]: According to the [zmscore docs](https://redis.io/commands/zmscore) ZMSCORE requires a minimum of one member. This either means that we either:

1. Validate that there's a non-zero length list (as above).
2. Validate that the item passed in is a string or a dictionary.

I'd rather we not overload the members argument, and instead always pass in a list. For the simple case, the list has one item (i.e ['foo']). Validating that we were passed a list I think is overkill, as the docs and the function signature are indicative of the requirement.

**Comment by chayim** [comment]
I think members should merely be required, and we shouldn't do a check on
it. It should be a list.  For what it's worth in the future I'm pretty pro
type hints, but we're not there.

--c

**Comment by 2014BDuck** [comment]
@chayim OK then. I am going to add a list-type default value now.

**Comment by 2014BDuck** [comment]
Ah sorry for the change. I am still thinking if we could just keep `None` as a default value.

So we can stick with the same validation:
```
# user pass an empty list / empty string / didn't pass member / etc
if not members:
    raise DataError()
```

And, use empty list as default won't stop communicating with redis server when user pass, for example "asdfasdf" or a dict contain something.

So the code is clean, and no risk for the default value being modified, same effect as we use empty list as default.

@chayim

**Comment by chayim** [comment]
For ZMSCORE, you have to have at a minimum one member
<https://redis.io/commands/zmscore>. and it's a list of members. We should
pass in a list, in order to not allow for overloadable constructs. A list
of length one, is still a list.

**Comment by 2014BDuck** [comment]
> For ZMSCORE, you have to have at a minimum one member <https://redis.io/commands/zmscore>. and it's a list of members. We should pass in a list, in order to not allow for overloadable constructs. A list of length one, is still a list.

Yes it requires a minimum one member so we don't need a default value, **as setting a `[]` as default won't have different behaviour as the current one**. Like I said it still won't pass the validation. And if user don't pass a list with length >= 1, the suggested changes won't help.

So the only thing it can help is to tell user that it should be a list. However user can notice this requirement in the function doc.

**Comment by chayim** [comment]
Two difference things are at play.  The first is that yes - we should
document accordingly.  Secondly, given that the only viable input is a
list, we should accept a list. If you want to validate the minimum length
of the list, to support the client use - I can support that! However, we
can't accept a default value of None, as the command doesn't support that.

**Comment by 2014BDuck** [comment]
Yes I see. 

So **currently** the default value is removed, which means user must pass `members` in. 
```
# user cannot do
r.zmscore("key")

# user can do:
r.zmscore("key", ["a", "b"])             # case 0
r.zmscore("key", "invalid_value")        # case 1
r.zmscore("key", [])                     # case 2
r.zmscore("key", {"invalide": "kv"})     # case 3
```

> we should accept a list.

**And with a `[]` as default value, user can still call this method with all 4 cases' input.** The `[]` is not helpful in limiting users' input.

Currently, this method can do:
- Users must send a `members` argument.

Can't:
- Limiting users to send a list input (The possible way is to validate by `isinstance`, which is suggested to change at the very beginning)

My point is, the default value won't affect any validation logic in the function body. No matter what default value we set, users can pass anything they want. So why not just don't set the default value and require an argument to be passed in. And we will have 2 different ways to handle the rest:
1. validate if it's a list with item: `if isinstance(members, list) and len(members) > 0:`.
2. A less solid validation: `if members:` and hand it over to Redis server. (Currently using)

I think either way we don't need the default value.

**Comment by chayim** [comment]
There are really only two cases present - the right items being specified, or not. The problem with validating the existence of a list, is that it doesn't prevent a user from passing in an invalid list - hence my being against that valiadation.  

Take as an example:

```
x = ['a', 'b']
y = ['a', 'b', 999, {'hello': 'world'}
z = ['a', None]
assert isinstance(x, list)
assert isinstance(y, list)
assert isinstance(z, list)
```

etc.

In this example, True will always be accurate, as everything is a list. I can appreciate the isinstance check on a list, but I wouldn't do anything beyond that. The use will receive a ResponseError should they provide invalid input to redis. Were we to sanitize the inputs in this function, we'd be responsible for raising a DataError (or to be fair an AttributeError) exception in all cases. I'd rather allow redis to do what it does and surface accordingly.

The following is a valid case, and I could understand generally doing this. However, I feel that [type hints](https://www.python.org/dev/peps/pep-0484/) are a better answer. 

```
x = ['some','list','of','things']
if not isinstance(x, list):
    raise AttributeError("x must be a list.")
```

**Comment by 2014BDuck** [comment]
OK, I am more clear about the responsibility of redis-py and redis server now. 

But I can't agree with using a mutable object as a default value. It's not recommended in any way and I never do this. Type hint could do this, but using `[]` as default value is just absolutely a different (and not secure) thing.

So for the function signature part: Use `None` as default(not preferred as we discussed) or just don't set a default value. Since user need to understand that they have to pass a list, I think we should keep the current code without a default value.

And for the body part, I've removed the `isinstance()` validation for a long time.  I think we could just be specific on the function signature now if you have a better idea.

**Comment by 2014BDuck** [comment]
By the way, we have warnings for unresolved reference on the `Commands` class (for the `execute_command` method) after the refactor in https://github.com/andymccurdy/redis-py/pull/1534 . Is it awarded and planed to fix?

**Comment by chayim** [comment]
Hey @2014BDuck I'll take a look at the warning in another issue - or happily accept a PR. As always, thanks for this contribution, it's great to see!

**Pull request merged**